src/main/java/net/fabricmc/mappingio/format/JarReader.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2021 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.mappingio.format;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemAlreadyExistsException;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Locale;
+
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+import net.fabricmc.mappingio.MappingVisitor;
+
+public class JarReader {
+	public static void read(Path path, MappingVisitor mappingVisitor) throws IOException {
+		mappingVisitor.visitNamespaces("jar", new ArrayList<String>());
+		processFile(path, null, new AnalyzingVisitor(mappingVisitor));
+	}
+
+	private static final class DirVisitor extends SimpleFileVisitor<Path> {
+		DirVisitor(AnalyzingVisitor visitor) {
+			this.visitor = visitor;
+		}
+
+		@Override
+		public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+			buffer = processFile(file, buffer, visitor);
```

Comment:
In jars, some directories aren't valid package names, like ones with `-`, also the META_INF and the version-specific class files.

Reply:
That doesn't handle say classes in doc-files or multi-release classes in META-INF

Reply:
Your process will yield invalid package names.

Reply:
I recommend skipping any class in any folder with - in name, as those are invalid java package names. However, I don't know how you want to deal with nested jars, and you might still consider jars in those directories.